### DISCUSSION TITLE
Netlify Sample App Testing Observations & Fixes

### DISCUSSION THREAD

**Discussion by kathirsvn** [stargate/stargate-mongoose#18] [General]
I have tested the netlify sample app with our latest driver and json API and made it work after below changes.

1. Skipped deleteMany calls in the app since it’s not implemented on the server side yet.
2. Skipped dropCollection calls (Is this the deleteCollection implemented on the server side? If yes, this is a work pending on the driver side.)
3. When the mongoose sends the updateOne call to the driver, it sends with a field called ‘session’ in the options which is not recognized by the server and throws back an error, so I had to remove the session field explicitly in the driver code on my local. Below is the sample request (from driver to the JSON API server). 
```
{
	"updateOne":
	{
		"filter":{"_id":"63effb9413d846a68462ae99","__v":0},
		"update":
		{
			"$set": {"items":[{"productId":"63effb9413d846a68462ae98","quantity":1}],"updatedAt":"2023-02-17T22:11:32.193Z"},
			"$inc":	{"__v":1}
		},
		"options": {"session":null}
	}
}
```
4. skipped ‘projection’ field in find and findOne for now, since server doesn’t recognize the field. Also, mongoose sends 'projection' as an object with ‘projection’ being one of the keys i.e. {"projection":{}}, this needs to be clarified with Val.

### **Steps to test the sample netlify app on local**
**Start the JSON API server on local**

- git clone git@github.com:stargate/jsonapi.git
- ./mvnw clean package -Dquarkus.container-image.build=true
- cd docker-compose
- ./start_dse_68_dev_mode.sh

**Package driver for local testing**

- git clone git@github.com:stargate/stargate-mongoose.git
- npm install
- npm run build
- npm pack
- This will create a file called stargate-mongoose-0.1.1.tgz

**Test the sample apps**

- git clone git@github.com:stargate/stargate-mongoose-sample-apps.git
- cp /path/to/stargate-mongoose-0.1.1.tgz stargate-mongoose-sample-apps/.
- cd stargate-mongoose-sample-apps/netlify-functions-ecommerce
- Edit package.json 'stargate-mongoose' dependency as `file:../stargate-mongoose-0.1.1.tgz`
- Create a keyspace called ‘netapp’ using swagger-ui
- set ASTRA_URI environment variable in the format: http://localhost:8180/v1/netapp?applicationToken={REPLACE_WITH_TOKEN}
- npm test
- README has options to start the app server

The above steps are for testing the sample app without npm publish and needs to be wrapped into a single downloadable docker file for any demo purposes.

**Comment by vkarpov15** [comment]
1. That's fine, we can use `deleteCollection()` as a workaround for now. Sample apps only use `deleteMany()` for clearing test data.
2. I implemeted `dropCollection` in #17 with https://github.com/stargate/stargate-mongoose/pull/17/commits/96e21f800069ec43581bf5e4c702816f7c7f6b76
3. This is a Mongoose issue, we'll fix that early next week: https://github.com/Automattic/mongoose/issues/13052
4. Partially a Mongoose issue: https://github.com/Automattic/mongoose/issues/13050 . However, Mongoose `save()` does use `projection` in one case: when there's no changes to save, Mongoose uses [`exists()`](https://github.com/Automattic/mongoose/blob/master/lib/model.js#L353-L370), which is a shorthand for [`findOne()` with `{ _id: 1 }`](https://github.com/Automattic/mongoose/blob/master/lib/model.js#L1174-L1177). We can work around this for a demo by simply omitting projections, but I think this means basic projection support is more important than I originally anticipated.

**Reply by kathirsvn** [reply]
1. Agreed on `deleteMany` we have a workaround for now.
2. `dropCollection` we are good now after this PR #17 
3. Thanks for creating an issue on the mongoose end as well, I also see the work around in the PR #17 and I think we are good on this for now.
4. Thanks again for the issue on the mongoose end as well to track and fix this. I agree on the `projection` requirement. We can discuss with @amorton next week on the basic `projection` support as required.

**Reply by amorton** [reply]
> 4\. Thanks again for the issue on the mongoose end as well to track and fix this. I agree on the `projection` requirement. We can discuss with @amorton next week on the basic `projection` support as required.

@kathirsvn  and @vkarpov15 yes you should be able to send an empty projection, or no `projection` clause. 

I have created this issue to support that https://github.com/stargate/jsonapi/issues/158
